### ISSUE TITLE
[Desktop] Allow users to add site exceptions for data that is cleared "on exit"

### ISSUE THREAD

**Issue by Brave-Matt** [brave/brave-browser#10493]
## Description 

As it is now, any data type that is checked in the `Clear browsing data --> [On exit]` will have _all data_ of that type removed upon closing the browser. While this is the intended function, we should implement a function similar to what Microsoft Edge does, where site exceptions can be added to a "Do not clear" list. Data on this list is ignored when the browser runs "clear on exit":

![image](https://user-images.githubusercontent.com/3299098/85903585-82d53680-b7bb-11ea-8677-61d71c6d3792.png)

This would help users maintain login data for sites they regularly visit and not have to re-authenticate every time data is erased. Additionally, this helps with support, since clearing browsing data often resolves issues with bad cache data -- having the ability to suggest users do this, but without sacrificing specific data they know they want would be very useful and appreciated by users

## Miscellaneous Information:

Original thread that bought this to my attention (again):
https://community.brave.com/t/delete-all-cookies-on-exit-except-allowed-list/137872/5

**Comment by newuser-web** [comment]
@Brave-Matt #10125 All bookmarks sites are usually important. So prioritising #10125 would be better.

**Comment by newuser-web** [comment]
@Brave-Matt #10125 All bookmarked sites are important. So prioritising #10125 would be better.

**Comment by newuser-web** [comment]
similar request #10125

**Comment by newuser-web** [comment]
@Brave-Matt Prioritise issue number 10125.

**Comment by irfaneru** [comment]
Dear sir what could I do if I lost my rewards bookmarks and settings also and password as the main character .
Using a mobile Android based brave browser

**Comment by bsclifton** [comment]
> Dear sir what could I do if I lost my rewards bookmarks and settings also and password as the main character .
> Using a mobile Android based brave browser

@irfaneru you're going to want to post on https://community.brave.com/ or search GitHub for a related issue. This issue is tracking a feature request for clearing data on exit

**Comment by RealExony** [comment]
This is such a simple yet crucial feature for a privacy focused browser that can still be used conveniently.

**Comment by sojusnik** [comment]
There's definitely demand for this very helpful feature, as it is constantly requested, as f.i.

https://community.brave.com/t/cookie-whitelist/82489
https://community.brave.com/t/clear-cookies-and-site-data-when-you-quit-brave-not-working-consistently/112651/3
https://community.brave.com/t/cookies-delete-in-all-sites-problem/128814
https://github.com/brave/brave-browser/issues/5089

**Comment by luke-shields** [comment]
This would really help a lot of people achieve better security, I'm sure more people would use the delete cookies on exit feature if you were able to add exceptions

**Comment by manolismih** [comment]
Please add this feature! Ecosia.org search engine asks for captcha verification every time i open the brave browser because te cookies are deleted on exit! Github asks for login etc etc. Thanks a lot for providing brave, it's about time we take privacy seriously!

**Comment by dfmarchlinski** [comment]
Definitely add this feature. The absence of this feature is the only thing stopping me from making Brave as the only browser I use.

**Comment by Brave-Matt** [comment]
+1 
https://community.brave.com/t/is-there-a-way-to-retain-a-list-of-specific-cookies-on-exit-when-clear-cookies-on-exit-is-enabled/266646/2

**Comment by bzdzz** [comment]
agreed, feature worth moving up the list of priority, I would like to have security of Clear All on exit with exception cookies in domains to keep logins I use frequently, thank you

**Comment by Brave-Matt** [comment]
+1 from Community:
https://community.brave.com/t/selecting-what-cookies-to-keep-and-delete-all-others/283635/2?u=mattches

**Comment by Brave-Matt** [comment]
+1 from Community:
https://community.brave.com/t/save-cookies-for-specific-websites-possible/288464

**Comment by aystic** [comment]
This feature is already present in Mozilla Firefox, MSEdge and I would love to see it in Brave browser as well. Its quite a helpful feature.

**Comment by Hooregi** [comment]
It really is a needed feature, I recently made the switch from firefox -> ungoogled-chromium -> brave
Didn't know this wasn't a feature to begin with. In the meantime I'm using cookie-autodelete, but I'd love for this to be implemented. Any ETA?

**Comment by joaopenteado** [comment]
I may be wrong, but I'm getting the feeling this feature is already implemented, albeit it's not as intuitive as it is on the other browsers. Like @Hooregi and many others, I have been using Cookie AutoDelete so far but decided to do a test in a different profile, and I've found that on `Brave v1.33.106` the **Customized behaviors** section does indeed override global cookie settings (as it should).

**Example 1.** If you block all cookies and add websites to the `Sites that can always use cookies` or `Always clear cookies when windows are closed` sections, those websites will indeed have the expected behavior and be able to use cookies as configured.

**Example 2.** If you check `Allow all cookies` or `Block third-party cookies`, enable `Clear cookies and site data when you close all windows` box and add websites to the `Sites that can always use cookies` section, those websites' cookies will persist between sessions.

**Comment by VegaDark** [comment]
> I may be wrong, but I'm getting the feeling this feature is already implemented, albeit it's not as intuitive as it is on the other browsers. Like @Hooregi and many others, I have been using Cookie AutoDelete so far but decided to do a test in a different profile, and I've found that on `Brave v1.33.106` the **Customized behaviors** section does indeed override global cookie settings (as it should).
> 
> **Example 1.** If you block all cookies and add websites to the `Sites that can always use cookies` or `Always clear cookies when windows are closed` sections, those websites will indeed have the expected behavior and be able to use cookies as configured.
> 
> **Example 2.** If you check `Allow all cookies` or `Block third-party cookies`, enable `Clear cookies and site data when you close all windows` box and add websites to the `Sites that can always use cookies` section, those websites' cookies will persist between sessions.

You are right, those are the settings. I don't know why the other people keeps saying there is no an actual setting to delete certain cookies and to keep others.

**Comment by sbucek** [comment]
You are wrong @joaopenteado.

**The goal is that  the "Always clear cookies when windows are closed" function always deletes all cookies except for sites specifically added as exceptions by user.**


-  **"Example 1"** requires to "block all cookies " which is impractical since some sites do not function without cookies. Manually adding those sites does not make sense either since not all of them must be kept after browser restart. 

- Tested **"Example 2"** in Brave v1.33.106, cookies got deleted even for sites added under "Sites that can always use cookies" after restarting Brave.

**Comment by joaopenteado** [comment]
> You are wrong @joaopenteado.
> 
> **The goal is that the "Always clear cookies when windows are closed" function always deletes all cookies except for sites specifically added as exceptions by user.**
> 
> * **"Example 1"** requires to "block all cookies " which is impractical since some sites do not function without cookies. Manually adding those sites does not make sense either since not all of them must be kept after browser restart.
> * Tested **"Example 2"** in Brave v1.33.106, cookies got deleted even for sites added under "Sites that can always use cookies" after restarting Brave.

Are you sure? I've just rebooted my machine and tested again and it seems to be working as intended.

**Comment by sbucek** [comment]
100% sure, "Sites that can always use cookies" does not override "Clear cookies and site data when you close all windows" as you state... Even after restarting my Linux machine...

**Comment by joaopenteado** [comment]
I've tested it on my Windows 11 and macOS machines and got the expected behavior without any issues. Unfortunately, I don't have a Linux machine at hand anymore to test it over there, but it might be some inconsistent behavior across different OSes or some extension conflict on your end (on my end I've made sure to use clean profiles for testing). It would be nice if we got a few more people to test this behavior on different OSes and post their findings here.

I think it might be worth considering reclassifying this issue as a bug fix rather than a new feature. We only need to make sure that the `Customized behaviors` for certain websites will always override the `Clear cookies and site data when you close all windows` and `Block all cookies` settings. The only enhancement that could be made would be maybe tweaking the UI to be little more similar to what is found on most other browsers, but that is very subjective "improvement".

**Comment by sbucek** [comment]
You're right, creating a new profile fixed it for me. Good job @joaopenteado ! Tweaking the UI would definitely be appreciated though.

**Comment by Hooregi** [comment]
So, I followed the advice to set up a new profile to test this and you all are absolutely right. It is a feature.

If I set `Clear cookies and site data when you close all windows` to `on` and add sites to `Sites that can always use cookies` those will not be deleted.

My problem came from setting `Clear browsing data -> On exit -> Cookies and other site data`, that's the one which **isn't** overwritten by the site exceptions. Thank you everyone!

**Comment by joeljose350** [comment]
> I may be wrong, but I'm getting the feeling this feature is already implemented, albeit it's not as intuitive as it is on the other browsers. Like @Hooregi and many others, I have been using Cookie AutoDelete so far but decided to do a test in a different profile, and I've found that on `Brave v1.33.106` the **Customized behaviors** section does indeed override global cookie settings (as it should).
> 
> **Example 1.** If you block all cookies and add websites to the `Sites that can always use cookies` or `Always clear cookies when windows are closed` sections, those websites will indeed have the expected behavior and be able to use cookies as configured.
> 
> **Example 2.** If you check `Allow all cookies` or `Block third-party cookies`, enable `Clear cookies and site data when you close all windows` box and add websites to the `Sites that can always use cookies` section, those websites' cookies will persist between sessions.

I can confirm that this works, but it is confusing and not user-friendly, in my opinion. Brave could do what Edge is doing, keeping a list of URLs whose data the user wants to preserve and deleting the data of every other website.

**Comment by Brave-Matt** [comment]
+1 from Community:
https://community.brave.com/t/deleting-cookies-on-exit-with-exceptions/359863/2

**Comment by EdwardLangdon** [comment]
No work on this still yet ?

**Comment by developomp** [comment]
> No work on this still yet ?

You can use [this method](https://github.com/brave/brave-browser/issues/10493#issuecomment-1002838700) for now.

**Comment by EdwardLangdon** [comment]
@developomp This doesn't work for me and see - https://www.reddit.com/r/brave_browser/comments/tvi0zu/comment/i3c0le/

**Comment by EdwardLangdon** [comment]
@developomp The workaround doesn't seem to work and see this comment - https://www.reddit.com/r/brave_browser/comments/tvi0zu/comment/i3c0le3/

**Comment by EdwardLangdon** [comment]
@developomp Thanks the method works.

**Comment by Kruln** [comment]
+1!!

**Comment by KOHPMC** [comment]
I think I understand what is happening now and why people are getting inconsistent results. There are two places to clear cookies on exit:

- Settings > clear browser data > switch to the "on exit" tab > "cookies and other site"

- Settings > privacy and security > cookies and other site data > clear cookies and site data when you close all windows.  It is a toggle. 

If you have the setting>clear browser data > on exit > "cookies and other site data" selected then it clears everything on exit -- including any sites you have in your "sites that can always use cookies".

To fix the issue, uncheck the field at: settings> clear browser data > on exit > "cookies and other site data".  Then go to:  settings > privacy and security > cookies and other site data > "clear cookies and site data when you close all windows". Here, keep the toggle set to "on".

This clears of the conflict between the two settings and keeps the cookies that are always set to on.

**Comment by onchov** [comment]
> > **Example 1.** If you block all cookies and add websites to the `Sites that can always use cookies` or `Always clear cookies when windows are closed` sections, those websites will indeed have the expected behavior and be able to use cookies as configured.
> 
> I can confirm that this works, but it is confusing and not user-friendly, in my opinion. Brave could do what Edge is doing, keeping a list of URLs whose data the user wants to preserve and deleting the data of every other website.

+1, couldn't put it better, using the opportunity to bump the issue.

**Comment by ASweatyPenguin-1** [comment]
This seems like a very basic feature to not implement.  This issue has been open a long time, a lot of other browsers have it, I'm considering using brave, but I need this feature.

**Comment by AManNeedsAMaid** [comment]
How is this still an issue?

How do you expect users to use Progressive Web Apps from within brave if you can't exclude certain sites from clearing cookies on exit?

**Comment by Brave-Matt** [comment]
+1 from Community:
https://community.brave.com/t/cannot-save-cookies-for-specific-websites-like-search-brave-com-if-delete-cookies-on-exit-is-on/496269/3

**Comment by port19x** [comment]
This is a really important feature for keeping some comfort with the increased privacy of clearing everything on exit

**Comment by CaptTom9** [comment]
I just tried to start using Brave. So far, pretty good. Until I noticed that there's no (straightforward) way to exclude domains from having cookies cleared on exit. Or, if the workaround referenced above does work, it's not at all clear that's the intended result, based on the wording and layout of the settings dialog.

I'm amazed at how long this appears to have been a problem with Brave. This is core stuff. All the other browsers do this. Having this outstanding for so long makes Brave look like a third-rate wannabe browser. I think this will be the end of my trial of Brave.

**Comment by chuzmunguz** [comment]
+100!!

I've been on the fence for several days now to move from Firefox to Brave, and this is such a basic feature and a major issue that seems to be around for so long, that I question the real motivation of brave developers and their commitment for the privacy-oriented community. The browser is full of crap-features like VPN, BAT, WEB3 stuff, ads, etc., but lack such a basic feature that allies privacy with convenience. If I could get extensions on Android, I would let this pass, but that's not even possible, so I can't get the Cookie AutoDelete extension to achieve what all other browsers can do natively. Unfortunately, I'm writing this in LibreWolf, and will stick with Mull browser on Android until the issue is fixed.

**Comment by AManNeedsAMaid** [comment]
> +100!!
> 
> I've been on the fence for several days now to move from Firefox to Brave, and this is such a basic feature and a major issue that seems to be around for so long, that I question the real motivation of brave developers and their commitment to the privacy-oriented community. The browser is full of crap-features like VPN, BAT, WEB3 stuff, ads, etc., but lack such a basic feature that allies privacy with convenience. If I could get extensions on Android, I would let this pass, but that's not even possible, so I can't get the Cookie AutoDelete extension to achieve what all other browsers can do natively. Unfortunately, I'm writing this on LibreWolf, and will stick with Mull browser on Android until the issue is fixed.

Agreed, especially lazy for a browser with its own customized settings page. Adding cryptocrap without addressing basic missing features is a terrible look.

**Comment by CaptTom9** [comment]
Well, I didn't give up on Brave after all. I've been testing the workaround:
`Settings / Privacy & Security / Cookies and other site data / Sites that can always use cookies
`
It seems that adding domains there does maintain their cookies between sessions. I'm still testing, and there are a few other quirks but at least the big hurtle _seems_ to be behind me - the need to clear all cookies on exit _except_ those sites I specify. I'm not sure why that's not explained and labeled better though.

**Comment by chuzmunguz** [comment]
I appreciate the tip on how to manage to get the exceptions on the desktop app, and maybe it can help other people, but I was already aware of it before I wrote my previous comment. The problem is that it seems that the Android app don't have the same option. 
I really would like to be able to sync browsing data on both desktop and mobile platforms (there has to be a balance between privacy and usability), but, at the present time, it looks like the behavior in the Android app is that **all** cookies are deleted when you select `Clear data on exit`, without an option to define exceptions.

I also tested the scenario:
`Clear data on exit` : `disabled`
`Settings / Brave Shields & privacy / Forget me when I close this site` :  `enabled`, with the exception added to the site Brave Shield while visiting it.


In theory, the end result for a single site would be the same for both scenarios. The problem is that with this method, some cookies persist between sessions, and that is not acceptable.

**Comment by aystic** [comment]
I am using the following settings on Brave desktop and they seem to work fine
- Go to `brave://settings/cookies` and toggle `Clear cookies and site data when you close all windows` option to be **ON** and add sites to the list `Sites that can always use cookies`
- Go to `brave://settings/clearBrowserData` then go to `On Exit` in the modal and **toggle on** the options `Browsing history`, `Download history` and `Cached images and files`

**Comment by CaptTom9** [comment]
To be clear, there are two settings which seem to apply here:

1) Settings / Privacy and Security / Clear Browsing Data / On Exit / Cookies and Other Site Data (check box)

2) Settings / Privacy and Security / Cookies and other site data / Clear cookies and site data when you close all windows (toggle button)

My next project will be to do some experiments with each setting, which means testing four different possibilities (both on, both off, one on, the other on.)

I'm wondering what "close all windows" really means. I don't usually run with more than one browser window open. That's what tabs are for. So for me, "on exit" means the same as "close all windows." I'm guessing I've misunderstood and there's another distinction Brave is making here. Having those two options in two different places also implies a different purpose.

Feel free to run the same tests and post what you find.

I'm primarily testing the desktop version. I don't like to sync settings with my mobile devices, so I won't be looking at that issue.

**Comment by Brave-Matt** [comment]
+1 from Community:
https://community.brave.com/t/customized-behavior-still-clears-every-time-i-close-and-restart-the-brave-browser/511649/3

**Comment by EriOldMan** [comment]
> I think I understand what is happening now and why people are getting inconsistent results. There are two places to clear cookies on exit:
> 
> * Settings > clear browser data > switch to the "on exit" tab > "cookies and other site"
> * Settings > privacy and security > cookies and other site data > clear cookies and site data when you close all windows.  It is a toggle.
> 
> If you have the setting>clear browser data > on exit > "cookies and other site data" selected then it clears everything on exit -- including any sites you have in your "sites that can always use cookies".
> 
> To fix the issue, uncheck the field at: settings> clear browser data > on exit > "cookies and other site data". Then go to: settings > privacy and security > cookies and other site data > "clear cookies and site data when you close all windows". Here, keep the toggle set to "on".
> 
> This clears of the conflict between the two settings and keeps the cookies that are always set to on.

I followed these steps but still had the issue. After trying everything I found out that you also need to uncheck: clear browser data > on exit > "Site and Sheilds Settings"

**Comment by sbucek** [comment]
@EriOldMan I have the same exact settings but still some sites get their cookies deleted despite being in the "Sites that can always use cookies" section. This feature not user friendly and completely unreliable as it is at the moment. It should be bulletproof.

**Comment by EriOldMan** [comment]
@sbucek 
That's odd. I agree but it's better than nothing for now.

**Comment by CaptTom9** [comment]
> This feature is not user friendly and completely unreliable as it is at the moment.

I couldn't agree more. I _think_ I've got it set OK now, but I'm not sure I'm clearing all the cookies I want. Sometimes sites seem to "remember" things I didn't expect them to. Using a browser shouldn't require in-depth analysis of what's being saved and what isn't. There should be clear, unambiguous options to clear all cookies on exit, and a clear way to identify exceptions in the same option setting dialog.

Oh, and when I went to try to set the same options on the Android version, nothing looked familiar. I gave up and returned to the browsers I'd been using before. This is really a shame. Brave could be a real star in the browser world!

**Comment by jlford** [comment]
+1 from community member, I'd really love to see this feature added, thank you!

**Comment by sbucek** [comment]
Did anyone test the **"Forgetful Browsing"**  browser feature  in Brave ( https://brave.com/privacy-updates/25-forgetful-browsing/ ) ?

Seems to be a much better alternative to  **"clean on exit"** as it removes site data already when closing the tab and can easily be configured for any specific site by activating the option in the navigation bar.  No need to enter URLs and stuff in the settings.

A clearer and more coherent integration between those 2 would be great though.

**Comment by IkelAtomig** [comment]
There is no option where I can edit Sites that can use cookies while as others are cleared by  clear cookies and site data when you close all windows.

Brave seems to have removed options. Kind of annoying as I cannot edit it or turn off so other sites can use cookies for login and making me re-login again and again.

**Comment by onchov** [comment]
> There is no option where I can edit Sites that can use cookies while as others are cleared by this :
> 
> `brave://settings/cookies` → `Clear cookies and site data when you close all windows`
> 
> Brave seems to have removed the option in a new update. Kind of annoying as I cannot edit it or turn off so other sites can use cookies for login, and making me re-login again and again.

I think it got moved to Settings > Privacy and security > Site and shields settings > Additional content settings > On-device site data, then you choose the middle option (Delete data sites that have been saved to your device when you close all windows) and define the exceptions below.

**Comment by daqnal** [comment]
@onchov is correct. I thought I was paranoid scrolling through this thread and not being able to find the toggles. Looks like it was changed in this issue #35415 , hence this from the release notes:
> Fixed out of sync cookie settings under brave://settings/cookies and brave://settings/shields.
This issue really needs to be revamped at a higher priority; this method for adding exceptions is very obscure, but the desire for it is not. Hopefully this is just an interim update and the team will prioritize it soon.

**Comment by IkelAtomig** [comment]
Damn. Thanks a lot. @onchov

**Comment by VGJUNKY** [comment]
This function works as others have said, you must untick "cookies and other site data" and "site and shield settings" in the clear browsing tab "on exit" brave://settings/clearBrowserData

then go to brave://settings/content/siteData and change the default behaviour to the second option "Delete data sites that have been saved to your device when you close all windows"

then add the sites you want to keep data for to the "Allowed to save data on your device"

Posting this only to make it simpler for people to figure out. No one else here has made it that easy to figure out. I think this should be closed or changed to a ui improvement request.

**Comment by CaptTom9** [comment]
Thank you, @VGJUNKY! This is much clearer.

Of course we're not always in this post to copy and paste the local URLs, above. To find the same settings in the UI, here is how to navigate through the Brave menus:

(1) Uncheck the "on exit" box:
- Settings
- Privacy and security
- Clear browsing data
- On Exit
- Cookies and other site data

(2) And the one I always have trouble finding, "Allow to save data on your device":
- Settings
- Privacy and security
- Site and Shields Settings
- Additional content settings
- On-device site data

**Comment by chris-1243** [comment]
> This function works as others have said, you must untick "cookies and other site data" and "site and shield settings" in the clear browsing tab "on exit" brave://settings/clearBrowserData
> 
> then go to brave://settings/content/siteData and change the default behaviour to the second option "Delete data sites that have been saved to your device when you close all windows"
> 
> then add the sites you want to keep data for to the "Allowed to save data on your device"
> 
> Posting this only to make it simpler for people to figure out. No one else here has made it actually simple to follow with all the steps. I think this should be closed or changed to a ui improvement request.

Thanks a lot for the explanation !!! An UI improvement would be useful and appreciated. I come from Firefox and it is way easier to fine tune this parameter.  You save me tons of search hours

**Comment by L1ttleFlyyy** [comment]
The initial suggestion by @VGJUNKY  have been working great so far. There is a tiny little button in address bar that allows you to add an exception very easily. And these per-site settings are even synced on sync chain.
Feel like just need to make some menu UI tweak to make this option more obvious to ordinary users.

**Comment by Brave-Matt** [comment]
+1 https://community.brave.com/t/bring-bag-ability-to-safe-specific-cookies-while-deleting-all-others-on-exit/615369/2

**Comment by gregforcey** [comment]
> There is a tiny little button in address bar that allows you to add an exception very easily.

Are you talking about the site settings button just the left of the URL. You select `Cookies and site data`? Or is there an easier way?

**Comment by amano-kenji** [comment]
We already have this feature, but accessing this feature requires a bit of learning curve. If someone didn't explain how to use the feature in this issue, I wouldn't have found the feature.

I think the issue is user experience. This feature is not obvious for most users.

At least, firefox makes this feature more obvious.

**Comment by Brave-Matt** [comment]
Still demand for this -- +1:
https://community.brave.app/t/how-to-keep-certain-sites-after-closing/643028

**Comment by HarimbolaSantatra** [comment]
For some reason, [this response](https://github.com/brave/brave-browser/issues/10493#issuecomment-2078572455) does not work for me.
One way or another, there's definitely UI problem.
It is not obvious where `brave://settings/content/siteData` is.
